### PR TITLE
[DiffusionGemma] Avoid recomputing token entropy in the stopping criteria

### PR DESCRIPTION
**Pull request by hsamrat007**
[CI](https://transformers-ci.lor-e.huggingface.cool/d/pytest-observability-by-pr/pytest-observability-branch?var-pr=47582)

# What does this PR do?

`Categorical(logits).entropy()` is computed twice per denoising step on the same tensor.

In `_denoising_step`, `processed_logits` goes to two consumers. `EntropyBoundSampler.accept_canvas` (called at `generation_diffusion_gemma.py:1050`) computes per-token entropy at lines 437-438 to decide which tokens to commit. `StableAndConfidentStoppingCriteria.__call__` (called at line 1065) computes the same entropy again at lines 533-534 to decide whether to stop.

This PR computes it once: `accept_canvas` returns the entropy it already calculated, and the stopping criteria accepts it as an optional argument.

For diffusiongemma-26B-A4B the logits are `[batch, 256, 262144]` fp32, which is 268 MiB per pass, so the second computation is not cheap.

## Cost

Both consumers are wrapped in `torch.compile` (lines 1248-1250 and 1258-1263), gated by `is_compiling` at lines 706-708, so the redundant cost depends on the path:

| path | redundant work per step |
|---|---|
| compiled (default) | ~268 MiB, since inductor fuses each side to roughly one pass |
| eager (`disable_compile=True`, or a non-compileable cache) | ~1.9 GiB, since `torch.distributions.Categorical` expands to ~7 passes (the constructor alone does a `logsumexp` plus a broadcast subtract into a fresh allocation) |

Measured on the **eager path**, on an NVIDIA GB10 (DGX Spark), bf16, fixed seed, 50 prompts per benchmark, patched vs unpatched interleaved per-prompt in one process. The runs pass no cache to `generate`, so it builds a `DynamicCache`, `is_compileable` is `False`, and the gate at lines 706-708 selects eager; a torch profile of the run confirms it (separate `logsumexp`/`softmax`/elementwise kernels rather than inductor-fused ones). Eager is also what makes the in-process A/B valid: there are no cached compiled bound methods on this path, so swapping the patch between arms takes effect immediately.

| benchmark | patched | unpatched | speedup | outputs |
|---|---|---|---|---|
| GSM8K (50) | 48/50 @ 75.2 tok/s | 48/50 @ 70.7 tok/s | +6.4% | token-identical 50/50 |
| HumanEval (50) | 46/50 @ 95.3 tok/s | 46/50 @ 90.0 tok/s | +5.9% | token-identical 50/50 |

Denoising step counts were also identical on all 100 prompts.

I have not benchmarked the compiled path. The redundant work there is ~7x smaller (one fused pass instead of ~7), so the expected gain is correspondingly smaller; the correctness verification below covers both paths.

## Correctness

No behavioral change. The entropy is still computed once by the same code on the same tensor; the second consumer now receives it instead of recomputing it. A caller that does not pass `token_entropy` falls back to computing it as before, so `StableAndConfidentStoppingCriteria` used standalone is unaffected.

On bit-exactness: in eager mode, both consumers ran the same `Categorical(...).entropy()` on the same tensor, and PyTorch reductions are run-to-run deterministic for a fixed shape, dtype and device, so the two results were already the same bits and reusing one changes nothing. Under inductor, the two entropies previously came from two separately compiled graphs, which could in principle pick different reduction strategies, so I verified token-identical output against `main` on both paths. The verification matters because the stopping rule is a threshold on `mean(token_entropy)`, and a last-ULP difference could flip a step count.

### The one case where reuse is not valid

When `finished_denoising.any()` is true, line 1061 rebuilds `processed_logits` from `self_conditioning_logits`:

```python
processed_logits = torch.where(
    finished_denoising[:, None, None], self_conditioning_logits, processed_logits
)
```

The sampler's entropy then describes a tensor that no longer exists, so reuse is dropped there and the stopping criteria recomputes as before. This is why `token_entropy = None` sits inside that branch.

## Why return the entropy instead of stashing it on the sampler

`_prepare_sampler` (line 698) constructs a fresh `EntropyBoundSampler` on every `generate()` call, but `_compile_functions` caches the compiled bound method on the model. From the second `generate()` call onward, the compiled function is still bound to the first sampler while `_denoising_step` holds the second, so a stash on `self` would be correct but would silently stop reusing anything after the first call. Returning the value avoids depending on instance identity. (`accepted_token_mask` works today only because `renoise_canvas` is stale-bound to the same old sampler; I'll flag that in a separate issue rather than change it here.)

The returned tensor is cloned in `_denoising_step` for the same reason `accepted_canvas` already is: it is a CUDA-graph output under `mode="reduce-overhead"` and is held across the `renoise_canvas` call. It is `[batch, canvas]` fp32, so the clone is negligible.

## Precedent

This follows two existing patterns in the generation code:

- `generation/stopping_criteria.py:54`: `StoppingCriteria.__call__(input_ids, scores)`. The AR stopping criteria receive the `scores` the decoding loop already computed rather than recomputing them. Passing loop-computed tensors into stopping criteria is the existing convention; recomputing was the outlier here.
- `generation/candidate_generator.py:44`: `CandidateGenerator.get_candidates` returns `(candidate_ids, candidate_logits)`, a generation component returning an auxiliary tensor it already computed so that the caller does not recompute it during assisted-generation verification. Same shape as `accept_canvas` returning `(accepted_canvas, token_entropy)`.

## Breaking change

`EntropyBoundSampler.accept_canvas` now returns `(accepted_canvas, token_entropy)` instead of just `accepted_canvas`. The three call sites in `tests/models/diffusion_gemma/` are updated. I could not find any other caller in the repo.

## Tests

`pytest tests/models/diffusion_gemma/`: 126 passed, 80 skipped, 0 failed. `make fix-repo` clean.

## Who can review?

@Cyrilvallez

### CODE REVIEW

**Comment by zucchini-nlp** [comment]
No code agents PRs pls as first-time contribs, and also we're not accepting any updates on DiffusionGemma, unless it is a severe bugfix. The model has been moved to `diffusers` library
